app/models/user.rb

```diff
@@ -125,6 +126,19 @@ def set_avatar_from_twitter
   rescue
   end
 
+  # Extract the information that is available after OmniAuth authentication.
+  def self.from_omniauth(auth, params)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.provider = auth.provider
```

Comment:
You don't need to add the `user.provider` or `user.uid` bits here – they'll automatically be created on line 131.